### PR TITLE
MM-17023: Plugin Marketplace

### PR DESCRIPTION
**Pull request by lieut-data**
#### Summary
Merging the feature branch for the plugins marketplace MVP.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-17023

#### Test Information
This PR was tested on: Chrome

### CODE REVIEW

**Comment by lieut-data** [comment]
Thanks for offering to help out, @jwilander!

File: src/reducers/entities/plugins.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    };
+}
+
+function marketplacePlugins(state = getInitialState(), action) {
+    switch (action.type) {
+    case PluginTypes.RECEIVED_MARKETPLACE_PLUGINS: {
+        return {
+            ...state,
+            data: action.data ? action.data : [],
+            serverError: null,
```
Comment (jwilander) [inline src/reducers/entities/plugins.js]: Not sure if we've stored the server error like this in the redux store before. Generally we just have the actions return the error and the component that called the action stores it in state. Not a big deal, and I'm 0/5, but thought I'd comment on the pattern difference.

File: src/reducers/entities/plugins.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        data: [],
+        serverError: null,
+    };
+}
+
+function marketplacePlugins(state = getInitialState(), action) {
+    switch (action.type) {
+    case PluginTypes.RECEIVED_MARKETPLACE_PLUGINS: {
+        return {
+            ...state,
```
Comment (jwilander) [inline src/reducers/entities/plugins.js]: nit: Seems unnecessary as all the state's fields are overridden anyways

Comment (marianunez) [reply src/reducers/entities/plugins.js]: Fixed

File: src/reducers/entities/plugins.js (hunk 1 above)
Comment (marianunez) [reply src/reducers/entities/plugins.js]: It is true that there are not a lot of cases that actually handle the [`onFailure`](https://github.com/mattermost/mattermost-redux/blob/ae8adedd4808d4d5f37f3be4f8d57767f28f0de6/src/actions/plugins.js#L13) case of the client function. However, there are a couple like this one: https://github.com/mattermost/mattermost-redux/blob/f80b0932d958a554e8eda8904573717b3b0ee48f/src/reducers/requests/users.js#L23 that do return the error in the state.

It gets kinda tricky to handle the return error directly from the call when using this pattern of the [`bindClientFunc`](https://github.com/mattermost/mattermost-redux/blob/ae8adedd4808d4d5f37f3be4f8d57767f28f0de6/src/actions/plugins.js#L10) which is what I was trying to follow as most of the actions in the repo do.

Comment (jwilander) [reply src/reducers/entities/plugins.js]: Thread resolved on the webapp PR

**Pull request merged**